pkg/bspatch/bspatch.go

```diff
@@ -100,128 +125,109 @@ func patchb(oldfile, patch []byte) ([]byte, error) {
 	//	from oldfile to x bytes from the diff block; copy y bytes from the
 	//	extra block; seek forwards in oldfile by z bytes".
 
-	// Read header
-	if n, err := f.Read(header); err != nil || n < 32 {
-		if err != nil {
-			return nil, fmt.Errorf("corrupt patch %v", err.Error())
-		}
-		return nil, fmt.Errorf("corrupt patch (n %v < 32)", n)
+	// Read the patch header
+	p := bytes.NewReader(patch)
+	n, err := p.Read(header)
+
+	if err != nil {
+		return newCorruptPatchError(err.Error())
 	}
+	if n < HeaderLen {
+		errmsg = fmt.Sprintf("short header read (n %v < %v)", n, HeaderLen)
+		return newCorruptPatchError(errmsg)
+	}
+
 	// Check for appropriate magic
 	if bytes.Compare(header[:8], []byte("BSDIFF40")) != 0 {
-		return nil, fmt.Errorf("corrupt patch (header BSDIFF40)")
+		return newCorruptPatchError("incorrect magic number (header BSDIFF40)")
 	}
 
 	// Read lengths from header
 	bzctrllen := offtin(header[8:])
 	bzdatalen := offtin(header[16:])
-	newsize = offtin(header[24:])
+	newsize := offtin(header[24:])
 
 	if bzctrllen < 0 || bzdatalen < 0 || newsize < 0 {
-		return nil, fmt.Errorf("corrupt patch (bzctrllen %v bzdatalen %v newsize %v)", bzctrllen, bzdatalen, newsize)
+		errmsg = fmt.Sprintf("negative length block(s) read from header (bzctrllen %v bzdatalen %v newsize %v)", bzctrllen, bzdatalen, newsize)
+		return newCorruptPatchError(errmsg)
 	}
 
 	// Close patch file and re-open it via libbzip2 at the right places
-	f = nil
-	cpf := bytes.NewReader(patch)
-	if _, err := cpf.Seek(32, io.SeekStart); err != nil {
-		return nil, err
+	p = nil
+	cpfbz2 := bytes.NewReader(patch)
+	if _, err := cpfbz2.Seek(int64(HeaderLen), io.SeekStart); err != nil {
+		return err
 	}
-	cpfbz2, err := bzip2.NewReader(cpf, nil)
+	cpf, err := bzip2.NewReader(cpfbz2, nil)
 	if err != nil {
-		return nil, err
+		return err
 	}
-	dpf := bytes.NewReader(patch)
-	if _, err := dpf.Seek(int64(32+bzctrllen), io.SeekStart); err != nil {
-		return nil, err
+	dpfbz2 := bytes.NewReader(patch)
+	if _, err := dpfbz2.Seek(int64(HeaderLen)+bzctrllen, io.SeekStart); err != nil {
+		return err
 	}
-	dpfbz2, err := bzip2.NewReader(dpf, nil)
+	dpf, err := bzip2.NewReader(dpfbz2, nil)
 	if err != nil {
-		return nil, err
+		return err
 	}
-	epf := bytes.NewReader(patch)
-	if _, err := epf.Seek(int64(32+bzctrllen+bzdatalen), io.SeekStart); err != nil {
-		return nil, err
+	epfbz2 := bytes.NewReader(patch)
+	if _, err := epfbz2.Seek(int64(HeaderLen)+bzctrllen+bzdatalen, io.SeekStart); err != nil {
+		return err
 	}
-	epfbz2, err := bzip2.NewReader(epf, nil)
+	epf, err := bzip2.NewReader(epfbz2, nil)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	pnew := make([]byte, newsize)
+	xbyteadd := newByteAddReader(dpf, oldf)
```

Comment:
What does `x` stand for? Just `byteadd` seems fine.

Reply:
meant to convey the zipping of the data block with the old file by x bytes per the notation in the file structure comment